internal/kubernetes/drainer.go

```
@@ -510,10 +530,97 @@ func (d *APICordonDrainer) GetPodsToDrain(node string, podStore PodStore) ([]*co
 		gracePeriod = *pod.Spec.TerminationGracePeriodSeconds
 	}
+	return gracePeriod
+}
+func (d *APICordonDrainer) evictWithKubernetesAPI(pod *core.Pod, abort <-chan struct{}) error {
+	gracePeriod:=d.getGracePeriod(pod)
+	return d.evictionSequence(pod, abort,
+		//eviction function
+		func() error {
+			return d.c.CoreV1().Pods(pod.GetNamespace()).Evict(&policy.Eviction{
+				ObjectMeta:    meta.ObjectMeta{Namespace: pod.GetNamespace(), Name: pod.GetName()},
+				DeleteOptions: &meta.DeleteOptions{GracePeriodSeconds: &gracePeriod},
+			})
+		},
+		//error handling function
+		func(err error) error {
+			// The eviction API returns 500 if a pod
+			// matches more than one pod disruption budgets.
+			// We cannot use apierrors.IsInternalError because Reason is not set, just Code and Message.
+			if statErr, ok := err.(apierrors.APIStatus); ok && statErr.Status().Code == 500 {
+				return OverlappingDisruptionBudgetsError{} // this one is typed because we match it to a failure cause
+			}
+			return err // unexpected (we're already catching 429 and 500), may be a client side error
+		},
+	)
+
+}
+
+func (d *APICordonDrainer) evictWithOperatorAPI(url string, pod *core.Pod, abort <-chan struct{}) error {
+	d.l.Info("using custom eviction endpoint", zap.String("pod", pod.Namespace+"/"+pod.Name), zap.String("endpoint", url))
+	gracePeriod:=d.getGracePeriod(pod)
+	maxRetryOn500:=4
+	return d.evictionSequence(pod, abort,
+		//eviction function
+		func() error {
+			evictionPayload := &policy.Eviction{
+				ObjectMeta:    meta.ObjectMeta{Namespace: pod.GetNamespace(), Name: pod.GetName()},
+				DeleteOptions: &meta.DeleteOptions{GracePeriodSeconds: &gracePeriod},
+			}
+
+			req, err := http.NewRequest("POST", url, GetEvictionJsonPayload(evictionPayload))
+			req.Header.Set("Content-Type", "application/json")
+
+			client := &http.Client{Timeout: 10 * time.Second}
+			resp, err := client.Do(req)
+			if err != nil || resp == nil {
```

When can err be nil and resp too?

---

I have not really check if this case exist (today), but I am using the `resp` pointer out if this `if` clause, this way I am sure to always be in the safe side (especially tomorrow if the case shows up).

---

I'd let it crash, because it should never happen.

> If the returned error is nil, the Response will contain a non-nil Body...

https://golang.org/pkg/net/http/#Client.Do

---

ok, done